### PR TITLE
Let statements in expressions.

### LINKED ISSUE #959: PIL: Implement `let` statement as expression
At expression level, `let`-statements should be allowed, together with blocks of statements.

This means, the following should be added to the alternatives for "expression":
`"{"` (LetStatement ";")* Expression "}"`

(Also the `IfStatement` and others should be extended so that we don't have to use double `{`.)

For now, the let statement requires a value (but see #960 for an extension).

During symbol resolution and evaluation, the defined variable is added to the `locals` just as we do for closures.

### PR DESCRIPTION
**Pull request by chriseth**
Allows braced blocks everywhere where expressions are expected.
The statements in those blocks can be `let x` or `let x = ...`.
The former declares a new witness column, the latter just binds a local variable.

fixes #959

### CODE REVIEW

**Comment by chriseth** [comment]
I'll do the `constr` part in a different pr.

File: ast/src/parsed/mod.rs [hunk 1]
```diff
@@ -658,6 +671,22 @@ impl<R> Children<Expression<R>> for IfExpression<R> {
     }
 }
 
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Clone, Serialize, Deserialize, JsonSchema)]
+pub struct LetStatementInsideBlock<Ref = NamespacedPolynomialReference> {
```
Comment (leonardoalt) [inline ast/src/parsed/mod.rs:675]: What's the difference between this and a normal let statement?

File: pil-analyzer/src/condenser.rs [hunk 2]
```diff
@@ -241,4 +293,39 @@ impl<'a, T: FieldElement> SymbolLookup<'a, T> for CondenserSymbols<'a> {
...
         Definitions(self.symbols).lookup_public_reference(name)
     }
+
+    fn new_witness_column(
+        &mut self,
+        name: &str,
+    ) -> Result<Arc<Value<'a, T>>, evaluator::EvalError> {
+        self.next_witness_id += 1;
+        let id = self.next_witness_id;
+        // Find an unused name.
```
Comment (leonardoalt) [inline pil-analyzer/src/condenser.rs:303]: Maybe extract this?

File: pil-analyzer/tests/condenser.rs [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    N.x_1 = N.y;
+"#;
+    let formatted = analyze_string::<GoldilocksField>(input).to_string();
+    assert_eq!(formatted, expected);
+}
+
+#[test]
+fn new_witness_column_name_clash() {
+    let input = r#"namespace N(16);
+    let new_wit = || { let x; x };
```
Comment (leonardoalt) [inline pil-analyzer/tests/condenser.rs:34]: lol

File: pil-analyzer/tests/condenser.rs [hunk 4]
```diff
@@ -0,0 +1,49 @@
+use powdr_number::GoldilocksField;
+use powdr_pil_analyzer::analyze_string;
+use test_log::test;
+
+use pretty_assertions::assert_eq;
+
+#[test]
+fn new_witness_column() {
+    let input = r#"namespace N(16);
+    let new_wit = || { let x; x };
```
Comment (leonardoalt) [inline pil-analyzer/tests/condenser.rs:10]: Can you also add some tests where this function adds constraints over the new column?

File: ast/src/parsed/mod.rs (hunk 1 above)
Comment (chriseth) [reply ast/src/parsed/mod.rs:675]: The other let statement is just an enum variant. The other difference is that here, we do not support types yet. Once add types, we can unify the two.

Comment (chriseth) [reply ast/src/parsed/mod.rs:675]: Actually I could unify them now and just not parse the type scheme.

Comment (chriseth) [reply ast/src/parsed/mod.rs:675]: ok, this gets a bit too complicated and I would like to defer that to later. The reason is that at parsing time, we have `Type<Expression<NamespacedPolynomialReference>>` and `Expression<NamespacedPolynomialReference>`, while after analysis we have `Type<u64>` and `Expression<Reference>`. Probably better to fix that by splitting the Expression type into multiple types.

File: pil-analyzer/tests/condenser.rs (hunk 4 above)
Comment (chriseth) [reply pil-analyzer/tests/condenser.rs:10]: Ah nice, that actually found a bug.

File: ast/src/parsed/mod.rs (hunk 1 above)
Comment (leonardoalt) [reply ast/src/parsed/mod.rs:675]: ok

File: pil-analyzer/tests/condenser.rs (hunk 4 above)
Comment (leonardoalt) [reply pil-analyzer/tests/condenser.rs:11]: I meant a case that adds constraints to the new column inside the function itself, like
```
let new_wit = || { let x; x = 2; x }
```

Or is that not possible atm?

Comment (chriseth) [reply pil-analyzer/tests/condenser.rs:11]: constraints can only be added in #1187

**Review by leonardoalt — APPROVED** [review]
please squash

**Review by leonardoalt — DISMISSED** [review]

**Review by leonardoalt — APPROVED** [review]

**Pull request merged**